src/analyz/mod.rs

```diff
@@ -95,37 +95,85 @@ fn vtable_descriptor_for_cast<'tcx>(
         return None;
     }
 
-    // Relevant code: rustc_codegen_ssa::base::unsized_info, and other functions in that file.
-    let old_pointee = match *old_ty.kind() {
-        ty::TyKind::Ref(_, ty, _) => ty,
-        ty::TyKind::RawPtr(ty, _) => ty,
-        _ => return None,
-    };
-    let new_pointee = match *new_ty.kind() {
-        ty::TyKind::Ref(_, ty, _) => ty,
-        ty::TyKind::RawPtr(ty, _) => ty,
-        _ => return None,
-    };
+    fn vtable_descriptor_for_ref_or_ptr_cast<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        old_pointee: ty::Ty<'tcx>,
+        new_pointee: ty::Ty<'tcx>,
+    ) -> Option<ty::PolyTraitRef<'tcx>> {
+        if !tcx.is_sized_raw(ty::TypingEnv::fully_monomorphized().as_query_input(old_pointee)) {
+            // We produce a vtable only for sized -> TyDynamic casts.
+            return None;
+        }
 
-    if !tcx.is_sized_raw(ty::TypingEnv::fully_monomorphized().as_query_input(old_pointee)) {
-        // We produce a vtable only for sized -> TyDynamic casts.
-        return None;
+        // Relevant code: rustc_codegen_ssa::meth::get_vtable
+        let trait_ref = match *new_pointee.kind() {
+            ty::TyKind::Dynamic(ref preds, _, _) =>
+                preds.principal().map(|pred| pred.with_self_ty(tcx, old_pointee)),
+            _ => return None,
+        };
```

Comment:
Eventually, for custom DST support, we'll need to update this to look into the last field of `TyKind::Adt`, so it returns the correct `TraitRef` (our stand-in for a vtable pointer, in this code) when `new_pointee` is e.g. `ArcInner<dyn Trait>`.  I'm not sure if we'll need a similar change elsewhere for array->slice unsizing - my guess is no, since mir-json just emits `Unsize` for those.